src/controller/mod.rs

```diff
@@ -1482,124 +1354,545 @@ impl Drop for ControllerInner {
+                let server = Http::new().bind(&addr, service.clone());
+                let server = match server {
+                    Ok(s) => s,
+                    Err(hyper::Error::Io(ref e))
+                        if e.kind() == ErrorKind::AddrInUse && addr.port() != 0 =>
+                    {
+                        Http::new()
+                            .bind(&SocketAddr::new(addr.ip(), 0), service)
+                            .unwrap()
+                    }
+                    Err(e) => panic!("{}", e),
+                };
+
+                let addr = server.local_addr().unwrap();
+                tx.send(addr).unwrap();
+                server.run_until(done_rx.map_err(|_| ())).unwrap();
+            })
+            .unwrap();
+
+        ServingThread {
+            addr: rx.recv().unwrap(),
+            join_handle,
+            stop: Box::new(done_tx),
+        }
+    }
+
+    /// Listen for messages from workers.
+    fn listen_internal(event_tx: Sender<ControlEvent>, addr: SocketAddr) -> ServingThread {
+        let mut done = Arc::new(());
+        let done2 = done.clone();
+        let mut pl: PollingLoop<CoordinationMessage> = PollingLoop::new(addr);
+        let addr = pl.get_listener_addr().unwrap();
+        let builder = thread::Builder::new().name("srv-int".to_owned());
+        let join_handle = builder
+            .spawn(move || {
+                pl.run_polling_loop(move |e| {
+                    if Arc::get_mut(&mut done).is_some() {
+                        return ProcessResult::StopPolling;
+                    }
+
+                    match e {
+                        PollEvent::ResumePolling(timeout) => {
+                            *timeout = Some(Duration::from_millis(100));
+                        }
+                        PollEvent::Process(msg) => {
+                            if event_tx.send(ControlEvent::ControllerMessage(msg)).is_err() {
+                                return ProcessResult::StopPolling;
+                            }
+                        }
+                        PollEvent::Timeout => {}
+                    }
+                    ProcessResult::KeepPolling
+                });
+            })
+            .unwrap();
+
+        ServingThread {
+            addr,
+            join_handle,
+            stop: Box::new(done2),
+        }
+    }
+
+    fn campaign(
+        event_tx: Sender<ControlEvent>,
+        authority: Arc<A>,
+        descriptor: ControllerDescriptor,
+        config: ControllerConfig,
+    ) -> JoinHandle<()> {
+        let descriptor = serde_json::to_vec(&descriptor).unwrap();
+        let builder = thread::Builder::new().name("srv-zk".to_owned());
+        builder
+            .spawn(move || {
+                loop {
+                    // become leader
+                    let current_epoch = authority.become_leader(descriptor.clone());
+                    let state = authority.read_modify_write(
+                        "/state",
```

Comment:
Is it deliberate that this doesn't write the nonce from the descriptor as part of the state? I don't quite follow how we check it if it's not stored in Zk. Or am I misunderstanding its role?